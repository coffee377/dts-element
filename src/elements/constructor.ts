import {Container} from '../collections';
import {any_type} from '../constants';
import {Element} from '../element';
import {emit_constructor_keyword} from '../helpers/emit-constructor-keyword';
import {emit_constructor_return} from '../helpers/emit-constructor-return';
import {emit_jsdoc} from '../helpers/emit-jsdoc';
import {emit_parameters} from '../helpers/emit-parameters';
import {Parameter} from './parameter';
import {Type} from './type';

export interface IConstructorRequiredParameters {
  parameters: Parameter[];
}

export interface IConstructorOptionalParameters {
  jsdoc: string;
  return: Type;
}

export class Constructor extends Element<IConstructorRequiredParameters, IConstructorOptionalParameters> {

  public get default_parameters(): IConstructorOptionalParameters {
    return {
      jsdoc: '',
      return: any_type,
    };
  }

  public _emit(container: Container): string {
    return `${this._emit_jsdoc()}${this._emit_raw(container)}`;
  }

  public _emit_jsdoc(): string {
    return emit_jsdoc(this.parameters.jsdoc);
  }

  public _emit_raw(container: Container): string {
    const keyword = emit_constructor_keyword(container);
    const return_type = emit_constructor_return(this.parameters.return, container);
    const parameters = emit_parameters(this.parameters.parameters, this);
    return `${keyword}(${parameters})${return_type};`;
  }

}
