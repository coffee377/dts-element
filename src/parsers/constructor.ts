import * as ts from 'typescript';
import {IGenericDeclaration} from '../declarations/generic-declaration';
import {IParameterDeclaration} from '../declarations/parameter-declaration';
import {create_class_member, IClassMember} from '../members/class-member';
import {parse_native} from '../parse';
import {create_constructor_type} from '../types/constructor-type';
import {has_kind, if_defined} from '../utils';

export const parse_constructor = (node: ts.ConstructorDeclaration): IClassMember =>
  create_class_member({
    owned: create_constructor_type({
      generics: if_defined(node.typeParameters, generics => generics.map(parse_native) as IGenericDeclaration[]),
      parameters: if_defined(node.parameters, parameters => parameters.map(parse_native) as IParameterDeclaration[]),
      return: if_defined(node.type, parse_native),
    }),
    readonly: has_kind(node.modifiers, ts.SyntaxKind.ReadonlyKeyword),
    optional: if_defined(node.questionToken, () => true),
  });
