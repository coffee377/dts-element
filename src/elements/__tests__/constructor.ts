jest.unmock('../constructor');

import {Constructor} from '../constructor';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new Constructor({parameters: []}).emit()).toMatchSnapshot();
  });
});
