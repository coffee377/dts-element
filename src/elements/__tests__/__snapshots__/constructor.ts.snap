// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-jsdoc][emit-constructor-keyword]([emit-parameters])[emit-constructor-return];"`;
